edx_solutions_api_integration/models.py

```diff
@@ -156,3 +159,187 @@ class APIUser(User):
+
+        if days_before_password_reset:
+            history = PasswordHistory.objects.filter(user=user).order_by('-time_set')
+            time_last_reset = None
+
+            if history:
+                # first element should be the last time we reset password
+                time_last_reset = history[0].time_set
+            else:
+                # no history, then let's take the date the user joined
+                time_last_reset = user.date_joined
+
+            now = timezone.now()
+
+            delta = now - time_last_reset
+
+            return delta.days >= days_before_password_reset
+
+        return False
+
+    @classmethod
+    def is_password_reset_too_soon(cls, user):
+        """
+        Verifies that the password is not getting reset too frequently
+        """
+        if not cls.is_password_reset_frequency_restricted():
+            return False
+
+        history = PasswordHistory.objects.filter(user=user).order_by('-time_set')
+
+        if not history:
+            return False
+
+        now = timezone.now()
+
+        delta = now - history[0].time_set
+
+        return delta.days < settings.ADVANCED_SECURITY_CONFIG['MIN_TIME_IN_DAYS_BETWEEN_ALLOWED_RESETS']
+
+    @classmethod
+    def is_allowable_password_reuse(cls, user, new_password):
+        """
+        Verifies that the password adheres to the reuse policies
+        """
+        if user.is_staff and cls.is_staff_password_reuse_restricted():
+            min_diff_passwords_required = \
+                settings.ADVANCED_SECURITY_CONFIG['MIN_DIFFERENT_STAFF_PASSWORDS_BEFORE_REUSE']
+        elif cls.is_student_password_reuse_restricted():
+            min_diff_passwords_required = \
+                settings.ADVANCED_SECURITY_CONFIG['MIN_DIFFERENT_STUDENT_PASSWORDS_BEFORE_REUSE']
+        else:
+            min_diff_passwords_required = 0
+
+        # just limit the result set to the number of different
+        # password we need
+        history = PasswordHistory.objects.filter(user=user).order_by('-time_set')[:min_diff_passwords_required]
+
+        for entry in history:
+
+            # be sure to re-use the same salt
+            # NOTE, how the salt is serialized in the password field is dependent on the algorithm
+            # in pbkdf2_sha256 [LMS] it's the 3rd element, in sha1 [unit tests] it's the 2nd element
+            hash_elements = entry.password.split('$')
+            algorithm = hash_elements[0]
+            if algorithm == 'pbkdf2_sha256':
+                hashed_password = make_password(new_password, hash_elements[2])
+            elif algorithm == 'sha1':
+                hashed_password = make_password(new_password, hash_elements[1])
+            else:
+                # This means we got something unexpected. We don't want to throw an exception, but
+                # log as an error and basically allow any password reuse
+                AUDIT_LOG.error('''
```

Comment:
It seem like import missing for AUDIT_LOG

Reply:
Okay i will add that.